x-pack/plugins/fleet/public/applications/fleet/hooks/use_url_pagination.ts

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useCallback, useEffect, useMemo } from 'react';
+import { useHistory, useLocation } from 'react-router-dom';
+import { useUrlParams } from './use_url_params';
+import { PAGE_SIZE_OPTIONS, Pagination, usePagination } from './use_pagination';
+
+type SetUrlPagination = (pagination: Pagination) => void;
+interface UrlPagination {
+  pagination: Pagination;
+  setPagination: SetUrlPagination;
+  pageSizeOptions: number[];
+}
+
+type UrlPaginationParams = Partial<Pagination>;
+
+/**
+ * Uses URL params for pagination and also persists those to the URL as they are updated
+ */
+export const useUrlPagination = (): UrlPagination => {
```

Comment:
I thought about changing the existing `usePagination()` to always persist the page/perPage to the URL, but that could have a large impact in testing to ensure it does not break anything. So for now, I went with a separate hook which could be merged and/or used if the use case calls for these parameters to be persisted in the URL.